src/main/java/com/faforever/api/map/MapService.java

```diff
@@ -0,0 +1,348 @@
+        || scenarioInfo.get("Configurations").get("standard").get("teams").get(1).get("name") == LuaValue.NIL
+        || scenarioInfo.get("Configurations").get("standard").get("teams").get(1).get("armies") == LuaValue.NIL
+        || !scenarioInfo.get("Configurations").get("standard").get("teams").get(1).get("name").tojstring().equals("FFA");
+  }
+
+  private void postProcessLuaFile(MapUploadData progressData) {
+    LuaValue scenarioInfo = progressData.getLuaScenarioInfo();
+    Optional<Map> mapEntity = mapRepository.findOneByDisplayName(scenarioInfo.get("name").toString());
+    if (mapEntity.isPresent()) {
+      if (mapEntity.get().getAuthor().getId() != progressData.getAuthorEntity().getId()) {
+        throw new ApiException(new Error(ErrorCode.MAP_NOT_ORIGINAL_AUTHOR, mapEntity.get().getDisplayName()));
+      }
+      int newVersion = scenarioInfo.get("map_version").toint();
+      if (mapEntity.get().getVersions().stream()
+          .anyMatch(mapVersion -> mapVersion.getVersion() == newVersion)) {
+        throw new ApiException(new Error(ErrorCode.MAP_VERSION_EXISTS, mapEntity.get().getDisplayName(), newVersion));
+      }
+      progressData.setMapEntity(mapEntity.get());
+    }
+  }
+
+  private void updateMapEntities(MapUploadData progressData) {
+    LuaValue scenarioInfo = progressData.getLuaScenarioInfo();
+    Map map = progressData.getMapEntity();
+    if (map == null) {
+      map = new Map();
+    }
+    map.setDisplayName(scenarioInfo.get("name").toString())
+        .setMapType(scenarioInfo.get("type").tojstring())
+        .setBattleType(scenarioInfo.get("Configurations").get("standard").get("teams").get(1).get("name").tojstring())
+        .setAuthor(progressData.getAuthorEntity());
+
+    LuaValue size = scenarioInfo.get("size");
+    MapVersion version = new MapVersion()
+        .setDescription(scenarioInfo.get("description").tojstring().replaceAll("<LOC .*?>", ""))
+        .setWidth(size.get(1).toint())
+        .setHeight(size.get(2).toint())
+        .setHidden(false)
+        .setRanked(progressData.isRanked())
+        .setMaxPlayers(scenarioInfo.get("Configurations").get("standard").get("teams").get(1).get("armies").length())
+        .setVersion(scenarioInfo.get("map_version").toint());
+    map.getVersions().add(version);
+    version.setMap(map);
+
+    progressData.setMapEntity(map);
+    progressData.setMapVersionEntity(version);
+
+    version.setFilename(progressData.getFinalZipName());
+    progressData.setFinalZipFile(Paths.get(
+        this.fafApiProperties.getMap().getFinalDirectory(),
+        progressData.getFinalZipName()));
+
+    if (Files.exists(progressData.getFinalZipFile())) {
+      throw new ApiException(new Error(ErrorCode.MAP_NAME_CONFLICT, progressData.getFinalZipName()));
+    }
+
+    // save entity to db to trigger validation
+    mapRepository.save(map);
+  }
+
+  @SneakyThrows
+  private void renameFolderNameAndAdaptLuaFiles(MapUploadData progressData) {
```

Comment:
Neither "Adapt" nor "update" is intuitive, time for JavaDoc on either or both of them (one may reference the other)